### PR TITLE
Incremental compile environment

### PR DESCRIPTION
**Pull request by jonludlam**
This is on top of the bugfix PR. The 'close environment' was causing many `Not_found` exceptions for various reasons, taking lots of time to resolve. This PR removes entirely the need to close an environment, thus fixing that class of bugs.

### CODE REVIEW

File: src/xref2/compile.ml [hunk 1]
```diff
@@ -194,64 +194,113 @@ and module_substitution env m =
...
+                  let rec find types rest =
+                    match rest with
+                    | Type (And, t) :: sgs -> find (t :: types) sgs
+                    | Type (_, _) :: _ -> types
+                    | Comment _ :: sgs -> find types sgs
+                    | _ -> types
+                  in
+                  let types = find [ t ] rest in
+                  List.fold_left add_to_env env types
+              | And | Nonrec -> env
```
[the remark this answers is not in the document]
Comment (jonludlam) [reply src/xref2/compile.ml:259]: The idea is that when we encounter a set of recursive modules we add them _all_ before handling any of the bodies - thus when we encounter an `and <module>` it's already in the environment so there's nothing to do.

File: src/xref2/type_of.ml [hunk 2]
```diff
@@ -10,23 +10,63 @@ let again = ref false
...
 
-and signature_items : Env.t -> Signature.t -> Signature.t =
- fun env s ->
+and signature_items : Env.t -> Signature.item list -> _ =
+ fun initial_env s ->
   let open Signature in
-  let items =
-    List.map
-      (fun item ->
+  let rec loop (items, env) xs =
```
[the remark this answers is not in the document]
Comment (jonludlam) [reply src/xref2/type_of.ml]: done, thanks!

File: src/xref2/type_of.ml [hunk 3]
```diff
@@ -10,23 +10,63 @@ let again = ref false
...
+    match xs with
+    | [] -> (List.rev items, env)
+    | item :: rest -> (
         match item with
-        | Module (r, m) -> Module (r, module_ env m)
-        | ModuleType mt -> ModuleType (module_type env mt)
-        | Include i -> Include (include_ env i)
-        | item -> item)
-      s.items
+        | Module (Nonrec, _) -> assert false
```
[the remark this answers is not in the document]
Comment (Julow) [reply src/xref2/type_of.ml:24]: Why do we need `Rec`, `Nonrec` and `Ordinary` ? The "ordinary" state for types would be `Rec` and for module would be `Nonrec`.

File: src/xref2/compile.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (Julow) [reply src/xref2/compile.ml:259]: Thanks.

File: src/xref2/type_of.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonludlam) [reply src/xref2/type_of.ml:24]: I think we could change the types, but let's do that in a separate PR. Keeping the same ones we already have limits the impact on the doc generator/loader in this PR.

**Pull request merged**